feat: AppArmor confinement for codejail-service

timmc-edx opened the pull request:
Allow codejail-service to actually run code (and run it securely) by giving it the needed confinement. Previously it would run but refuse to execute code since it would detect the insecure environment; now, the startup safety checks pass and the code-exec endpoint works as expected.

- Add an AppArmor profile with fairly strict rules. It needs to be thoroughly vetted and to have exceptions added before it can be used in production, but it's be fine for devstack. Some parts are based on the existing edxapp apparmor config without careful review.
- Apply the profile to the codejail service in docker-compose.
- Add Django configs for codejail service.
- Add documentation for installing the profile so that it is available for use on the dev's machine.

Also:

- Add configuration and documentation for edxapp to actually call the codejail service, disabled by default. (Will later want to make this default to true, once the service is working properly.)
- Update image name in docker-compose to follow rename in https://github.com/edx/public-dockerfiles/pull/102

Currently edxapp gets an error back from codejail-service, and then isn't able to read that error; separate work in the app repo will be needed to fix those. (The first issue relates to python_path, and the other to not returning globals_dict when there's an emsg.) But the integration is working otherwise.

----

I've completed each of the following or determined they are not applicable:

- [x] Made a plan to communicate any major developer interface changes (or N/A)

clintonb commented:
@edx/docker-devstack-working-group please review
FYI @edx/learner @edx/devops

timmc-edx committed:
feat: AppArmor confinement for codejail-service

Allow codejail-service to actually run code (and run it securely) by
giving it the needed confinement. Previously it would run but refuse to
execute code since it would detect the insecure environment; now, the
startup safety checks pass and the code-exec endpoint works as expected.

- Add an AppArmor profile with fairly strict rules. It needs to be
  thoroughly vetted and to have exceptions added before it can be used in
  production, but it's be fine for devstack. Some parts are based on the
  existing edxapp apparmor config without careful review.
- Apply the profile to the codejail service in docker-compose.
- Add Django configs for codejail service.
- Add documentation for installing the profile so that it is available for
  use on the dev's machine.

Also:

- Add configuration and documentation for edxapp to actually call the
  codejail service, disabled by default. (Will later want to make this
  default to true, once the service is working properly.)
- Update image name in docker-compose to follow rename in
  https://github.com/edx/public-dockerfiles/pull/102

Currently edxapp gets an error back from codejail-service, and then isn't
able to read that error; separate work in the app repo will be needed to
fix those. (The first issue relates to python_path, and the other to not
returning globals_dict when there's an emsg.) But the integration is
working otherwise.

timmc-edx commented on py_configuration_files/lms.py:
```diff
@@ -554,6 +554,12 @@ def should_show_debug_toolbar(request):  # lint-amnesty, pylint: disable=missing
+# Disabled by default since codejail service needs to be configured
+# and started separately. See docs/codejail.rst for details.
+ENABLE_CODEJAIL_REST_SERVICE = False
+CODE_JAIL_REST_SERVICE_HOST = "http://edx.devstack.codejail:8080"
```
Author's note: I had thought the gunicorn port wouldn't matter and I could just set it to a conventional value, but I forgot it would be used by inter-service communication in devstack. This makes for an odd mismatch with the rest of the ports seen in devstack. Maybe it's fine, though.

robrap replied:
You could leave a code comment:
```
The port does not match the pattern used for other devstack services because...
```

robrap replied:
@timmc-edx: Again, if you thought the note https://github.com/edx/devstack/pull/109#discussion_r1953596473 was worthwhile in the PR, why not add a code comment that helps clarify?

timmc-edx replied:
Oh yeah, forgot about this one... I was trying to figure out if there was a better place to document it, but I'll just add a comment.

timmc-edx commented on codejail.profile:
```diff
@@ -0,0 +1,75 @@
+
+    # Allow executing this binary, but force a transition to the specified
+    # profile (and scrub the environment).
+    /sandbox/venv/bin/python Cx -> child,
```
Author's note: Make it explicit that changes here need to be coordinated with the Dockerfile (and Django settings).

timmc-edx commented on codejail.profile:
```diff
@@ -0,0 +1,75 @@
+# production, but for now it is only intended to be used in devstack.
+
+
+#include <tunables/global>
```
Author's note: Add brief comment to explain (once I understand it).

timmc-edx commented on codejail.profile:
```diff
@@ -0,0 +1,75 @@
+
+# Declare ABI version explicitly to ensure that confinement is
+# actually applied appropriately on newer Ubuntu.
+abi <abi/3.0>,
```
Author's note: Add link to discussion of why this was needed.

timmc-edx commented on codejail.profile:
```diff
@@ -0,0 +1,75 @@
+# the outer one is going to help. But there may be some small value in
+# defense-in-depth, as it's possible that a bug in the child (sandbox)
+# profile isn't present in the outer one.
+profile codejail_service flags=(attach_disconnected,mediate_deleted) {
```
Author's note: Check which of these flags we actually need -- they're pretty standard but we need to understand them before we include them.

timmc-edx commented on codejail.profile:
```diff
@@ -0,0 +1,75 @@
+# defense-in-depth, as it's possible that a bug in the child (sandbox)
+# profile isn't present in the outer one.
+profile codejail_service flags=(attach_disconnected,mediate_deleted) {
+    #include <abstractions/base>
```
Author's note: Add brief comment to explain (once I understand it).

timmc-edx replied:
Also note that this is not a comment.

timmc-edx commented on codejail.profile:
```diff
@@ -0,0 +1,75 @@
+    signal (send) set=(kill) peer=codejail_service//child,
+
+    # Allow executing this binary, but force a transition to the specified
+    # profile (and scrub the environment).
```
Note here that this is what "Cx" means, specifically.

timmc-edx commented on py_configuration_files/cms.py:
```diff
@@ -312,6 +312,13 @@ def should_show_debug_toolbar(request):  # lint-amnesty, pylint: disable=missing
+
+# Disabled by default since codejail service needs to be configured
+# and started separately. See docs/codejail.rst for details.
+ENABLE_CODEJAIL_REST_SERVICE = False
```
```suggestion
#ENABLE_CODEJAIL_REST_SERVICE = True
```

timmc-edx committed:
fixup! Switch to no-# include; add comments

We can use either `#include` or `include` and it might be less confusing
to use the one that doesn't look like a comment.

I've added comments to some directives that I now understand better.

timmc-edx committed:
fixup! Remove attach_disconnected and add more comments

Just about every sample policy on the web includes attach_disconnected but
the manpage describes it as a debugging tool that is not safe for general
use.

timmc-edx committed:
fixup! add explanation for abi

timmc-edx committed:
fixup! add note about subprofile path

timmc-edx committed:
fixup! Make it easier to enable (rather than copying the default)

robrap commented on docs/codejail.rst:
```diff
@@ -0,0 +1,43 @@
+
+While AppArmor provides good protection, a sandbox escape could still be possible due to misconfiguration or bugs in AppArmor. For defense in depth, we're setting up a dedicated `codejail service <https://github.com/openedx/codejail-service>`__ that will perform code execution for edxapp and which will allow further isolation.
+
+The default edxapp codejail defaults to unsafe, direct execution of Python code, and this remains true in devstack. We don't even have a way to run on-host codejail securely in devstack. In constrast, the codejail service refuses to run if codejail has not been configured properly, and we've included a way to run it in devstack.
```
[opinion] I'd leave most of this for the openedx/codejail-service docs. It's going to change even more if the old behaviour is DEPRed (including removal). Here, I'd write more from the perspective that we have and want a codejail-service, and here is what is needed to work with it.

This is non-blocking. It is an opinion to minimize future doc updates as time passes.

Note: I could imagine an ADR that captures this in the codejail-service, if that doesn't already exist, and you could point to it if you feel it is worth it, without surrounding it with much redundant text.

timmc-edx replied:
Yeah, I'll go ahead and trim this down, and save what I wrote for when I write the docs for the app repo.

robrap commented on docs/codejail.rst:
```diff
@@ -0,0 +1,43 @@
+
+In order to run the codejail devstack component:
+
+1. Install AppArmor: ``sudo apt install apparmor``
```
- Is this directly on my local machine (Mac, etc.)? Could you make that more explicit?
- Once I'm clear, I can run through this as well.

timmc-edx replied:
Yes, it's meant to be run on the host machine. Although now I'm realizing that I'm not actually sure where you're supposed to run it on a Mac! It's a Linux command, and won't work for you.

How does this docker stuff all work on Mac, anyhow? Maybe there's a Linux VM that all the docker stuff runs inside? If so, this would need to be run in that Linux VM.

robrap replied:
Maybe this will help us? https://github.com/abiosoft/colima/discussions/1019

timmc-edx replied:
I don't see anything promising in there. My understanding is that Colima sets up an Ubuntu-based Linux VM and runs docker inside of that, and that to get this working on a Mac you'd have to access that VM directly and install apparmor (and this profile) inside of the VM. I see references on the web to `colima ssh`, which might help?

This is definitely going to be harder to run on a Mac.

robrap commented on docs/codejail.rst:
```diff
@@ -0,0 +1,43 @@
+
+Any time you update the profile, you'll need to re-run the command to apply the profile.
+
+The profile file contains the directive ``profile codejail_service``. That defines the name of the profile when it is installed into the kernel. In order to change that name, you must first remove the profile **under the old name**, then install a new profile under the new name. To remove a profile, use the ``--remove`` action instead of the ``-replace`` action: : ``sudo apparmor_parser --remove  ...
```
Is there supposed to be a relationship between the name `codejail.profile` and the earlier `profile codejail_service`? Can you better explain the connection?

timmc-edx replied:
Nope, filename is unrelated -- could be named anything. `profile codejail_service` is the important part. It tells apparmor to install a profile of this name in the OS.

robrap replied:
I started the review with this doc, and I asked this before I had my bearings. I didn't realize that `codejail.profile` was a file under review as part of this PR. Maybe you could note earlier on where to find the file `codejail.profile`, and we can adjust the doc when and if it moves?

timmc-edx replied:
What if I change this to `./codejail.profile`? That would help hint that it's a file in the repo.

robrap commented on docs/codejail.rst:
```diff
@@ -0,0 +1,43 @@
+
+Changes to the AppArmor profile must be coordinated with changes to the Dockerfile, as they need to agree on filesystem paths.
+
+Any time you update the profile, you'll need to re-run the command to apply the profile.
```
I was wondering how to do this. Is the next paragraph the answer? If so, maybe either start the next paragraph with "To re-run the command to apply the profile...", or just make this one paragraph that goes together. Or, any other way to clarify what how you to this. Thanks.

timmc-edx replied:
This was supposed to mean that the command to install the profile in the first place and the command to update it are one and the same. (It's the one with `--replace`.) It's a kind of confusing situation. I'll try to clarify the docs.

(Actually, I think you can use `--add` instead of `--replace`, but then every time afterwards you need to use `--replace`, which... I'd rather just use one command for both!)

robrap commented on docs/codejail.rst:
```diff
@@ -0,0 +1,43 @@
+
+The profile file contains the directive ``profile codejail_service``. That defines the name of the profile when it is installed into the kernel. In order to change that name, you must first remove the profile **under the old name**, then install a new profile under the new name. To remove a profile, use the ``--remove`` action instead of the ``-replace`` action: : ``sudo apparmor_parser --remove  ...
+
+The profile name must also agree with the relevant ``security_opt`` line in devstack's ``docker-compose.yml``.
```
I'm not sure exactly what this means.
1. Is this whole section about changing the name meant to be instructions for a permanent name change/improvement, or a useful development technique? If so, maybe you could explain why and when you'd want to change the name.
2. It sounds like there might be certain steps to take when changing a name. For example, removing the old profile name. Do you need to do anything to add the new name, or does remove do everything you want? Do you then need to update the `security_opt` line? If so, this would be better written: "Next, you must update the relevant `security_opt` line...`. If this is not what you meant at all, maybe this is enough to show you my confusion with this line? I'm not clear when and if you are asking me to change something, and which side to change.

timmc-edx replied:
It's basically supposed to be a warning that changing the name is not simple. I should probably replace this with something like "avoid changing the name, and here's why", and people can look up the process elsewhere if they need it. I'm hoping we can stick with this name and not change it.

robrap commented on codejail.profile:
```diff
@@ -0,0 +1,100 @@
+abi <abi/3.0>,
+
+# This outer profile applies to the entire container, and isn't as
+# important. If the sandbox profile doesn't work, it's not likely that
```
Maybe: "... isn't as important as the inner child profile."

Ah, I like discussing our inner child.

timmc-edx replied:
You know, I copied the name "child" from a pattern I'd seen elsewhere, but it would be more descriptive to name the inner profile "sandbox". What do you think?

robrap replied:
1. I made my first comment more clear with a suggestion.
```suggestion
# important as the inner child profile. If the sandbox profile doesn't work, it's not likely that
```
2. We've moved the possible "child" profile renaming discussion to https://github.com/edx/devstack/pull/109/files#r1960732376.

timmc-edx replied:
Added a change similar to this.

robrap commented on codejail.profile:
```diff
@@ -0,0 +1,100 @@
+# the outer one is going to help. But there may be some small value in
+# defense-in-depth, as it's possible that a bug in the child (sandbox)
+# profile isn't present in the outer one.
+profile codejail_service flags=(mediate_deleted) {
```
1. [proposal] I think using the name `codejail_service_profile` will make this more clear throughout other code and docs for those that are less familiar.
2. I wanted to know what `mediate_deleted` is for before I got to the answer. Maybe add something like: "See the child profile below for an explanation of the `mediate_deleted` flag."

timmc-edx replied:
I'd really like to avoid having the word "profile" in the profile name. It would be like having "variable" in a variable name.

Yeah, I can make a forward reference to the later comment.

robrap replied:
I'm going to make one more push for the name change. 

First, there is a reason that most of your english comments refer to "child profile", because the word "child" isn't meaningful enough on its own.

Second, I'm more concerned with the uses of the profile name than the declaration. For example:
```
signal (send) set=(kill) peer=codejail_service//child,
```

For those that don't know this syntax well, this is not as clear as:
```
signal (send) set=(kill) peer=codejail_service//child_profile,
```

Additionally, if you decide to rename "child" to "sandbox", the following gets even more ambiguous for those that aren't expert in this syntax:
```
signal (send) set=(kill) peer=codejail_service//sandbox,
```

timmc-edx replied:
The comment there currently says "Allow sending a kill signal to the sandbox ..." -- what if I change that to "Allow sending a kill signal to the `child` subprofile ..."? This would indicate what the `//` syntax means, and use its name explicitly.

(I still might also want to rename that inner profile. Maybe `jailed`?)

robrap replied:
I guess you really don't like the idea of using `_profile`. 😉 

Regarding possible renames, what about `codejail_sandbox`?

timmc-edx replied:
Yup, very much don't want it, if at all possible.

`codejail_sandbox` sounds fine!

robrap commented on codejail.profile:
```diff
@@ -0,0 +1,100 @@
+    # Allow access to a variety of commonly needed, generally safe things
+    # (such as reading /dev/random, free memory, etc.)
+    #
+    # Manpage: "Includes files that should be readable and writable in all profiles."
```
Is this from `man apparmor`, or other? 

I think it would be useful to note this somewhere, if you think these are the best docs. I'm inferring this from the install, but the only command I saw was `apparmor_parser`, so maybe making this explicit here and/or in the codejail.rst. 

Note: I know you wanted to limit redundancy with codejail-service docs, so you can decide when to refer to them and add updates to those docs instead. Thanks.

timmc-edx replied:
Actually `man apparmor.d` -- I'll include a link to https://manpages.ubuntu.com/manpages/noble/man5/apparmor.d.5.html at the top of the file.

robrap commented on docs/codejail.rst:
```diff
@@ -0,0 +1,43 @@
+
+Both LMS and CMS can run Python code submitted by instructors and learners in order to implement custom Python-graded problems. By default this involves running the code on the same host as edxapp itself. Ordinarily this would be quite dangerous, but we use a sandboxing library called `codejail <https://github.com/openedx/codejail>`__ in order to confine the code execution in terms of disk and ne ...
+
+While AppArmor provides good protection, a sandbox escape could still be possible due to misconfiguration or bugs in AppArmor. For defense in depth, we're setting up a dedicated `codejail service <https://github.com/openedx/codejail-service>`__ that will perform code execution for edxapp and which will allow further isolation.
```
It may make sense to either change this to point to the README: https://github.com/openedx/codejail-service/blob/main/README.rst, and/or refer to reading those docs, rather than just noting the service.

robrap commented on codejail.profile:
```diff
@@ -0,0 +1,100 @@
+        # includes the sandbox's copy of Python as well as any
+        # dependencies that have been installed for inclusion in
+        # sandboxes.
+        /sandbox/venv/** rm,
```
When I first read this, I got confused with the `rm` command. Maybe add a comment of "r - read, m - memory (or whatever this actually is"?
Note: I think rw is less confusing below, and having a single comment here would be good enough for removing confusion. I don't think we need to comment all the lines below in this way.

robrap commented on codejail.profile:
```diff
@@ -0,0 +1,100 @@
+    # apparmor will continue to make policy decisions in cases where a confined
+    # executable has a handle to a file's inode even after the file is removed
+    # from the filesystem.
+    profile child flags=(mediate_deleted) {
```
[proposal] Again, using the name `child_profile` will make this more clear throughout other code and docs for those that are less familiar.

robrap replied:
Proposal denied.

robrap commented on py_configuration_files/codejail.py:
```diff
@@ -1,3 +1,28 @@
+CODEJAIL_ENABLED = True
+
+CODE_JAIL = {
+    # These values are coordinated with the Dockerfile and AppArmor profile
```
Maybe something like this would help people find these references?
```suggestion
    # These values are coordinated with the Dockerfile (edx/public-dockerfiles) and AppArmor profile (codejail.profile)
```

timmc-edx committed:
fixup! Deny outbound connections; clean up capabilities

- Document what netlink is and why it's needed
- Allow TCP connections, but not establishing outbound ones
- Split apart capability lines into two blocks for easier commenting

timmc-edx committed:
fixup! Shrink docs for codejail setup (will move to service repo docs)

timmc-edx committed:
fixup! Clarify instructions

- `--add` vs `--replace` and when to use them
- TODO around Mac
- Hints that codejail.profile is a file provided in the repo

timmc-edx committed:
fixup! Clarify discussion of profile renaming

timmc-edx committed:
fixup! Use name of subprofile rather than a description of it

timmc-edx committed:
fixup! Link to manpage

timmc-edx committed:
fixup! Explain `m` access mode

timmc-edx committed:
fixup! More explicit locations for sandbox path coordination

robrap commented on codejail.profile:
```diff
@@ -8,12 +8,17 @@
 # production, but for now it is only intended to be used in devstack.
 
+
+
```
Just making sure you wanted 3 new lines. I'm good with whatever you want.

timmc-edx replied:
Yeah, just wanted some separation between the header and the next comment. (It was confusing me when I went to edit it.)

robrap commented on docs/codejail.rst:
```diff
@@ -6,34 +6,34 @@ The ``codejail`` devstack component (codejail-service) requires some additional
 Configuration
 *************
 
+**TODO**: These instructions are for Linux. We need some instructions for how to do this on Mac; it probably involves accessing the Linux VM that docker is run inside of.
```
How about something like this?
```suggestion
These instructions are for Linux only. Additional research would be required to create instructions for a Mac, which probably involves accessing the Linux VM that docker is run inside of.
```

robrap commented on docs/codejail.rst:
```diff
@@ -6,34 +6,34 @@ The ``codejail`` devstack component (codejail-service) requires some additional
-Any time you update the profile, you'll need to re-run the command to apply the profile.
-
-The profile file contains the directive ``profile codejail_service``. That defines the name of the profile when it is installed into the kernel. In order to change that name, you must first remove the profile **under the old name**, then install a new profile under the new name. To remove a profile, use the ``--remove`` action instead of the ``-replace`` action: : ``sudo apparmor_parser --remove  ...
+Any time you update the profile, you'll need to update the profile in your OS as well: ``sudo apparmor_parser --replace -W ./codejail.profile``
```
Something like this?
```suggestion
Any time you update the profile file, you'll need to update the profile in your OS as well: ``sudo apparmor_parser --replace -W ./codejail.profile``
```

timmc-edx committed:
fixup! Comment on port mismatch

timmc-edx committed:
fixup! Rename child profile to codejail_sandbox

Just a better (more descriptive) name, and has parallels to
codejail_service profile name.

timmc-edx committed:
fixup! Apply suggestion re: Mac TODO

Co-authored-by: Robert Raposa <rraposa@edx.org>

timmc-edx committed:
fixup! Add "file" for clarity

Co-authored-by: Robert Raposa <rraposa@edx.org>

timmc-edx committed:
fixup! Clarify what "important" is relative to
